components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java

```diff
@@ -311,23 +311,88 @@ public void deleteUser(String userId) throws NotFoundException, CharonException
                 requiredAttributes);
     }
 
-    private List<Object> listUsers(Map<String, Boolean> requiredAttributes) throws CharonException {
+    private List<Object> listUsers(Map<String, Boolean> requiredAttributes, int offset, int limit, String sortBy,
+            String sortOrder, String domainName) throws CharonException {
 
-        ClaimMapping[] coreClaims;
-        ClaimMapping[] userClaims;
-        ClaimMapping[] extensionClaims = null;
         List<Object> users = new ArrayList<>();
-        //0th index is to store total number of results;
+        //0th index is to store total number of results.
         users.add(0);
+
+        String[] userNames;
+
+        if (StringUtils.isNotEmpty(domainName)) { // When domain is not null or not empty.
+            if ((isPaginatedUserStoreAvailable()) && (carbonUM instanceof PaginatedUserStoreManager)) {
+                if (limit <= 0) {
+                    limit = getMaxLimit();
+                }
```

Comment:
Having this limit here will break backward compatibility, when listing users as previously it returned all users.
As there is a property to switch to old behavior that's bit fine yet, someone might not be able to use the new implementation at all. I think it would be better to have some thing to decide whether to get all or to adhere to the max limit

Reply:
Legacy APIs were not supported to any of the query params. Therefore Whatever the query params comes through the request, it retrieve all the users across all user stores. Here we set the limit to maxLimit if there is a domain in the request. Therefore this won't break backward compatibility.